### PR TITLE
mtx: add events commands

### PR DESCRIPTION
**Pull request by DeadByElpy**

### CODE REVIEW

File: src/mtx/utils/CommandBinaryBuffer.ts [hunk 1]
```diff
@@ -1552,6 +1573,81 @@ CommandBinaryBuffer.prototype.setExtendedCurrentValues2 = function ( parameters:
...
 
+CommandBinaryBuffer.prototype.getEvent = function (): IEvent {
+    const data: IEvent = {
+        hours: this.getUint8(),
+        minutes: this.getUint8(),
+        seconds: this.getUint8(),
+        event: this.getUint8()
+    };
+    const {event} = data;
+
```
Comment (DarkPark) [inline src/mtx/utils/CommandBinaryBuffer.ts:1584]:
```suggestion
```

File: src/mtx/commands/downlink/getEvents.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+import * as types from '../../types.js';
+import * as command from '../../utils/command.js';
+import * as accessLevels from '../../constants/accessLevels.js';
+import CommandBinaryBuffer, {ICommandBinaryBuffer} from '../../utils/CommandBinaryBuffer.js';
+
+
+interface IGetEventsParameters {
+    date: types.IDate,
+
```
Comment (DarkPark) [inline src/mtx/commands/downlink/getEvents.ts:38]:
```suggestion
```

not needed if there are no docs here

File: src/mtx/commands/uplink/getEvents.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
+/**
+ * Uplink command to get device events.
```
Comment (DarkPark) [inline src/mtx/commands/uplink/getEvents.ts:2]: it should probably match the downlink command description

File: src/mtx/commands/uplink/getEvents.ts [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+interface IGetCriticalEventResponseParameters {
+    date: types.IDate,
+    eventsNumber: types.TUint8,
+    events: Array<IEvent>
+}
+
+// date + event number byte
+const BODY_WITHOUT_EVENTS_SIZE = 3 + 1;
+const EVENT_SIZE = 4;
+export const id: types.TCommandId = 0x33;
```
Comment (DarkPark) [inline src/mtx/commands/uplink/getEvents.ts:76]:
```suggestion

export const id: types.TCommandId = 0x33;
```

?

File: src/mtx/commands/uplink/getEvents.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+ * Decode command parameters.
+ *
+ * @param bytes - command body bytes
+ * @returns decoded parameters
+ */
+export const fromBytes = ( bytes: types.TBytes ): IGetCriticalEventResponseParameters => {
+    if ( bytes.length > maxSize ) {
+        throw new Error(`Wrong buffer size: ${bytes.length}.`);
+    }
+    console.log(bytes);
```
Comment (DarkPark) [inline src/mtx/commands/uplink/getEvents.ts:142]:
```suggestion
```

File: src/mtx/utils/CommandBinaryBuffer.ts [hunk 6]
```diff
@@ -711,22 +711,40 @@ export interface IPackedEnergiesWithType {
...
+    TIME_CORRECT: boolean
+}
+
+export interface IEvent {
+    hours: types.TUint8,
+    minutes: types.TUint8,
+    seconds: types.TUint8,
+    event: types.TUint8,
+    power?: Array<types.TUint8>,
+    newDate?: {
```
Comment (DarkPark) [inline src/mtx/utils/CommandBinaryBuffer.ts:738]:
```suggestion
    newDate?: IDateTime
```

?

File: src/mtx/commands/uplink/getCriticalEvent.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+
+/**
+ * Encode command parameters.
+ *
+ * @param parameters - command parameters
+ * @returns full message (header with body)
+ */
+export const toBytes = ( parameters: IGetCriticalEventResponseParameters ): types.TBytes => (
+    command.toBytes(id, [
```
Comment (DarkPark) [inline src/mtx/commands/uplink/getCriticalEvent.ts:147]:
```suggestion
    const {event, index, date, count} = parameters;

    command.toBytes(id, [
```

?

File: src/mtx/commands/uplink/getEventsCounters.ts [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    remoteParametersChange: types.TUint16,
+    powerOff: types.TUint16,
+    restart: types.TUint16,
+    setClock: types.TUint16
+}
+
+
+const COMMAND_BODY_SIZE = 14;
+const OLD_COMMAND_BODY_SIZE = 20;
+export const id: types.TCommandId = 0x34;
```
Comment (DarkPark) [inline src/mtx/commands/uplink/getEventsCounters.ts:53]:
```suggestion

export const id: types.TCommandId = 0x34;
```

File: src/mtx/utils/CommandBinaryBuffer.ts (hunk 1 above)
Comment (DeadByElpy) [reply src/mtx/utils/CommandBinaryBuffer.ts]: fixed

File: src/mtx/commands/downlink/getEvents.ts (hunk 2 above)
Comment (DeadByElpy) [reply src/mtx/commands/downlink/getEvents.ts]: fixed

File: src/mtx/commands/uplink/getEvents.ts (hunk 3 above)
Comment (DeadByElpy) [reply src/mtx/commands/uplink/getEvents.ts]: fixed

File: src/mtx/commands/uplink/getEvents.ts (hunk 4 above)
Comment (DeadByElpy) [reply src/mtx/commands/uplink/getEvents.ts:77]: fixed

File: src/mtx/utils/CommandBinaryBuffer.ts (hunk 6 above)
Comment (DeadByElpy) [reply src/mtx/utils/CommandBinaryBuffer.ts]: fixed

File: src/mtx/commands/uplink/getCriticalEvent.ts (hunk 7 above)
Comment (DeadByElpy) [reply src/mtx/commands/uplink/getCriticalEvent.ts]: fixed

File: src/mtx/commands/uplink/getEventsCounters.ts (hunk 8 above)
Comment (DeadByElpy) [reply src/mtx/commands/uplink/getEventsCounters.ts:54]: fixed

**Review by DarkPark — APPROVED** [review]

**Pull request merged**